### ISSUE TITLE
The `useAsObservableSource` is causing React warning

### ISSUE THREAD

**Issue by SidKwok** [mobxjs/mobx-react-lite#274]
# ✋✋✋ Issue was revived based on https://github.com/mobxjs/mobx-react-lite/issues/274#issuecomment-652958242

The following comment is left here for clarity but doesn't reflect an actual issue.

---

## Intended outcome

Not trigger warning.

## Actual outcome

warning from react:
```log
index.js:1 Warning: Cannot update a component (`wrappedComponent`) while rendering a different component (`wrappedComponent`). To locate the bad setState() call inside `wrappedComponent`, follow the stack trace as described in https://fb.me/setstate-in-render
    in wrappedComponent (created by Context.Consumer)
    in Route (at App.tsx:48)
    in Switch (at App.tsx:47)
    in div (at App.tsx:46)
    in App (at App.tsx:60)
    in Router (created by HashRouter)
    in HashRouter (at App.tsx:59)
    in Unknown (at src/index.tsx:7)
```

## How to reproduce the issue
Try to run this code
```jsx
import React from 'react';
import { observer, useLocalStore } from 'mobx-react-lite';
import { useHistory, Route, Switch, HashRouter } from 'react-router-dom';

const context = React.createContext(null);

const Com = observer(() => {
  const store = React.useContext(context) as any;
  return (
    <div>
      <p>{store.str}</p>
      <button onClick={store.go}>
        test
      </button>
    </div>
  );
});

const Page = observer(() => {
  const history = useHistory();
  const storeSource = {
    history,
    search: history.location.search
  };
  const store = useLocalStore(source => ({
    get str() {
      return source.search
    },
    go() {
      source.history.replace(
        '/test?a=1'
      );
    }
  }), storeSource);
  return (
    <context.Provider value={store as any}>
      <div>
        <Com />
      </div>
    </context.Provider>
  );
});

function App() {
  return (
    <div className="App">
      <Switch>
        <Route
          path="/test"
          component={Page}
          exact
        />
      </Switch>
    </div>
  );
}

export default () => (
  <HashRouter>
    <App />
  </HashRouter>
);

``

## Versions

react: 16.13.1
react-dom: 16.13.1
react-router-dom: 5.1.2
mobx: 5.15.4
mobx-react-lite: 2.0.4

**Comment by FredyC** [comment]
Can you please use CodeSandbox to showcase the issue? It would easier for us to investigate more closely.

**Comment by SidKwok** [comment]
https://codesandbox.io/s/misty-wave-tkycp?file=/src/App.js
![image](https://user-images.githubusercontent.com/13268747/78852490-0a2dcc00-7a4f-11ea-93fb-d5c6b02e0ffc.png)

@FredyC click test and you will see the warning

**Comment by nulladdict** [comment]
I managed to strip down example to bare React + mobx-react-lite [sandbox link](https://codesandbox.io/s/serene-ganguly-1z2qd?file=/src/App.js)
Clicking increment first time indeed produces the warning

I'm pretty sure this happens due to the assignment during render in [this line](https://github.com/mobxjs/mobx-react-lite/blob/master/src/useAsObservableSource.ts#L27) which triggers reaction and eventually setState

I tried moving this assignment into the `useEffect` which does seem to work, this version runs without warnings:

```ts
React.useEffect(() => {
    runInAction(() => {
        Object.assign(res, current)
    })
})
```

However this increases render count and breaks some tests. I don't have enough inside knowledge about `react-mobx-lite` to understand if this behaviour is good or bad and might need consulting @FredyC 

I've created a [draft PR](https://github.com/mobxjs/mobx-react-lite/pull/280/) to try to fix this issue. Also while running tests for it I found a potential bug in `useAsObservableSourceInternal` implementation, which I'll describe in the PR

**Comment by FredyC** [comment]
To be honest, I am not sure either. This new flashy warning is causing havoc all around and it's more confusing than helpful in my opinion.

> I'm pretty sure this happens due to the assignment during render in [this line](https://github.com/mobxjs/mobx-react-lite/blob/master/src/useAsObservableSource.ts#L27) which triggers reaction and eventually setState

That sounds safe to me, even if observable changes during render phase, the reaction and `forceUpdate` are still happening a bit later inside the `useEffect`. There is no `setState` happening during the render phase afaik.

It's definitely curious why your _fix_ helped to remove that warning. Either it's some coincidence or there is actually something it. My expertise isn't that great either.

**Comment by FredyC** [comment]
I totally missed a reproduction from @SidKwok before, sorry about that. But looking into that, there is something smelly in here.

```jsx
const Parent = observer(() => {
  const [value, setValue] = useState({ current: 0 });
  const store = useLocalStore(
    source => ({
      get current() {
        return source.current;
      },
      increment() {
        setValue(prev => ({ current: prev.current + 1 }));
      }
    }),
    value
  );
  return <Child store={store} />;
});
```

First of all, why are you combining `useState` and `useLocalStore`? That feels like unnecessary complication without any benefit to it. However, what's more important, why do you call `setValue` inside the `increment` action. That's conceptually all wrong. Actions should modify observables, not the outside state value. This is most likely the source of that warning.

**Comment by nulladdict** [comment]
Also it is curious that the warning shows up only for the *first* click but not for any subsequent clicks

Regarding `useAsObservableSourceInternal`, I can submit a test case where source argument for `useLocalStore` changes from `{}` to `undefined` which causes `useLocalStore` to break in a different PR

**Comment by FredyC** [comment]
> Regarding `useAsObservableSourceInternal`, I can submit a test case where source argument for `useLocalStore` changes from `{}` to `undefined` which causes `useLocalStore` to break in a different PR

Well, that's more of the bug in user code. It's built to always expect the same structure. I guess we can do DEV warning, but it's definitely a code smell.

**Comment by nulladdict** [comment]
Also I do agree mutating something outside of observable scope which needs to observed for some reason is conceptually wrong

**Comment by FredyC** [comment]
@SidKwok closing the issue as it seems that your pattern is simply wrong. I will reopen if you have something else to add here.

**Comment by SidKwok** [comment]
> Wait, I totally missed a reproduction from @SidKwok before, sorry about that. But looking into that, there is something smelly in here.

@FredyC 
The case you shown is not my reproduction. My case is combining `history.replace` and `useLocalStore`. You can look into this link: https://codesandbox.io/s/misty-wave-tkycp?file=/src/App.js

**Comment by SidKwok** [comment]
@FredyC 
If the pattern is wrong, how can the state `search` produced by  `history` be observed in this case? Is there a best practice?

**Comment by FredyC** [comment]
You are right there is something strange to your use case. I stripped down your example to remove any "weird patterns" and the issue can still be reproduced.

https://codesandbox.io/s/quirky-taussig-qfwyc?file=/src/App.js

It does really seem that `useAsObservableSource` mutation in the render phase is problematic here.

@mweststrate Do you have some ideas here? There was attempted PR https://github.com/mobxjs/mobx-react-lite/pull/280 to run `useAsObservableSource` inside the `useEffect` but it breaks some tests, mainly about an increased number of re-renders. What's more important that I am failing to see what's so special about this use case that it causes that warning.

**Comment by FredyC** [comment]
Digging further and I removed react-router from the equation and the issue can be still reproduced even more reliably after every page reload (because the empty string is initial value).

https://codesandbox.io/s/festive-currying-498ny?file=/src/App.js

It almost seems that MobX does something special when an empty string is passed to `observable`. But it doesn't make any sense 😵.

**Comment by nulladdict** [comment]
> @FredyC
> If the pattern is wrong, how can the state `search` produced by `history` be observed in this case? Is there a best practice?

I think the general approach is to `useLocation` which subscribes to location changes and should re-render if search is changed

**Comment by FredyC** [comment]
@nulladdict That's pretty much irrelevant, the `useLocation` is just a different export of the same context. It might make more sense code-wise, but doesn't help anything actually.

**Comment by nulladdict** [comment]
What I meant is that it’s probably better to not observe state managed by something else, just like in my `useState` example, and just use that state directly

Also empty string probably has nothing to do with a warning, since it’s still reproduces with a number in my example. I believe the area of the problem is somewhere in initialization and first updating of observable

**Comment by FredyC** [comment]
So Michel had a look at it and the conclusion is that it seems like a false negative for that warning. There is technically nothing wrong with the pattern I've outlined in [previous sandbox](https://codesandbox.io/s/quirky-taussig-qfwyc?file=/src/App.js).

If we would go through modifying the `useAsObservableSource` and adding extra `useEffect`, we get extra re-renders for other people. That's a bad idea just for an edge case warning.

If the warning bothers you, it would be better to avoid MobX in this case because there doesn't seem to be an added value for you. At least in this convoluted example. I assume your real code is more complex, but try to rethink the approach. Basically what @nulladdict said applies here. Consume `useLocation` (or whatever hook) directly. If you want your own abstraction, make a custom hook.

The Golden rule is that you should pick the right tool for the job, not to try force single tool for every job.

**Comment by SidKwok** [comment]
Sorry about the jet lag between me and you guys.

> What I meant is that it’s probably better to not observe state managed by something else, just like in my useState example, and just use that state directly

Consider case like this:
```jsx
const { search } = useLocation();
const store = useLocalStore(source => ({
  get paramOne() {
    retrun source.search + 'xx';
  },
  paramTwo: 'paramTwo'
}), { search })

// make things reactive
useEffect(() => autorun(async () => {
  // get users from server whenever `search` or `paramTwo` change
  const user = await getUser({
    paramOne: store.paramOne
    paramTwo: store.paramTwo
  });
}), [])

// without autorun, I can still achieve it with useEffect
useEffect(() => {
  const paramOne = search + 'xx';
  // get users...
}, [search, store.paramTwo]);

// but it's a little bit tricky, right?
// what if it needs paramThree, Four, Five, which is also produced by `search`?

```

Luckily, it sounds like the warning is just some noise. But the pattern seems fine.

**Comment by SidKwok** [comment]
> If the warning bothers you, it would be better to avoid MobX in this case because there doesn't seem to be an added value for you. 

@FredyC 

The warning doesn't bother me. But I would like to make sure that there is nothing actually wrong in this case. If the warning is just some noises, that's fine. But if there is some hidden errors with this pattern, I may try to rethink the approach. It seems ok in my running website, for now.

**Comment by FredyC** [comment]
@SidKwok Yea, the warning is most likely false-positive given everything seems to work just fine. In production, it won't even happen and slow down anything.

However, there is definitely a risk that this pattern might cause problems with Concurrent mode and the warning is the way to get prepared for it. Totally ignoring it isn't probably the best to approach.

If Concurrent mode is a concern for you, it might be even worth opening an issue with React so someone more educated can identify if there is an actual problem. Or even trying to modify the sandbox to see if it works with a currently experimental version of React and Concurrent mode enabled.

For your example above, I think it's kinda overengineered. You can basically do a custom hook like the following if you want some sort of derivation. MobX won't give you much benefit here. In the contrary, it would complicate your code with `autorun/reaction` in components that are [recommended to be avoided](https://github.com/mobxjs/mobx-react/issues/772#issuecomment-526949917).

```js
const { search } = useLocation();
return { paramOne: search + 'xx' }
```

**Comment by SidKwok** [comment]
@FredyC 
Thanks. It helps a lot!

**Comment by olee** [comment]
@FredyC actually there **is** a huge problem with the logic in useAsObservableSource and this issue should definitely be reopened: 

The following code will cause the "Cannot update a component..." error because in `ObserverParent`, the reactions from useAsObservableSource will indirectly call setState (observer -> useForceUpdate -> setState) in `ObserverChild` to trigger a rerender, before the parent finished rendering ([CodeSandbox](https://codesandbox.io/s/staging-lake-6l45l?file=/src/App.js)).

```jsx
export default function App() {
  const [value, setValue] = React.useState("");
  return (
    <div>
      <div>Type something here</div>
      <input value={value} onChange={e => setValue(e.target.value)} />
      <ObserverParent value={value} />
    </div>
  );
}

export const ObserverParent = observer(props => {
  console.log("Render ObserverParent");
  const source = useAsObservableSource(props);
  return <ObserverChild source={source} />;
});

export const ObserverChild = observer(props => {
  console.log("Render ObserverChild");
  return <div>Value = {props.source.value}</div>;
});
```

As of now I don't see and fix around this other than using a construct like the following ([CodeSandbox](https://codesandbox.io/s/spring-dream-46wmr?file=/src/App.js:115-797)):

```jsx
export default function App() {
  const [value, setValue] = React.useState("");
  return (
    <div>
      <div>Type something here</div>
      <input value={value} onChange={e => setValue(e.target.value)} />
      <ObserverParent value={value} />
    </div>
  );
}

export const ObserverParent = observer(props => {
  console.log("Render ObserverParent");
  const [source] = React.useState(() => observable(props));
  React.useEffect(action(() => {
    Object.assign(source, props);
  }));
  return <ObserverChild source={source} />;
});

export const ObserverChild = observer(props => {
  console.log("Render ObserverChild");
  return <div>Value = {props.source.value}</div>;
});
```

**Comment by olee** [comment]
@FredyC could you perhaps allow globally overriding default useForceUpdate hook used in useObserver instead of just per single invocation?
So in [this line](https://github.com/mobxjs/mobx-react-lite/blob/ef91cca9d5fbb09ceaca286d2e21e9549cd19801/src/useObserver.ts#L45) instead of
```ts
    const wantedForceUpdateHook = options.useForceUpdate || useForceUpdate
```
do something like
```ts
    const wantedForceUpdateHook = options.useForceUpdate || defaultUseForceUpdate
```
where defaultUseForceUpdate is a variable that can be updated by something like `setDefaultForceUpdateHook(useAsyncForceUpdate)`

**Comment by FredyC** [comment]
If you have an idea on how to fix the problem, please do create a PR. I don't see a way. I don't follow how is your suggestion for global `useForceUpdate` going to work given you always need to dispatch for that particular component (from `useState`).

**Comment by olee** [comment]
The idea is to dispatch asynchronously **after** the current rendering process finished (basically something like `setTimeout(() => forceUpdate(), 0)`).
This would prevent components being updated during the rendering of another component.
You can check my example code in the [CodeSandbox](https://codesandbox.io/s/mobx-react-lite-issue-274-broken-6l45l?file=/src/App.js:2566-3284)

**Comment by FredyC** [comment]
I see. Ok, please try to submit a PR with tests passing. You can integrate it inside `useObserver`, no need to expose `forceUpdate` globally. However, I am kinda afraid of memory leaks. Especially with React Concurrent mode, it will be hell to properly know which forceUpdates are still live.

**Comment by olee** [comment]
Yeah, the issue for me right now is that if I would like to apply this fix for the whole project, I would have to replace all imports of `observer` with a custom implementation.
Instead what I would rather do is being able to set the default hook which should be used for `options.useForceUpdate` across all observer components.

As for memory leaks / concurrent mode: Yeah - I think as well that there will be problems, but I'm pretty sure those will also happen with the current implementation.
As a matter of fact, the current implementation **is** broken as you can see in the example code.

I will write up a PR which should at least allow users to circumvent these issues.
Other than that I think the only real solution is to never use useAsObservableSource in components but instead pass everything you need as observable sources instead of props.

**Comment by olee** [comment]
While this allows mobx-react-lite users to easily fix this issue across a project, I think we should probably open a new issue to think about a general solution (like making an asynchronous variant of useForceUpdate the default implementation).

Right now I don't see any other solution to the problem and when I think about it, it makes sense that modifications to the observables should only cause rerenders __after__ the current render finished.

**Comment by olee** [comment]
Example for a simple fix when the PR is merged and released:

Provide useAsyncForceUpdate
```js
import React from "react";
import { unstable_batchedUpdates } from "react-dom";

// List of tasks to run on next tick
let asyncForceUpdateQueue = [];

function runAsyncForceUpdateQueue() {
  const tasks = asyncForceUpdateQueue;
  asyncForceUpdateQueue = [];
  unstable_batchedUpdates(() => {
    for (const task of tasks) {
      task();
    }
  });
}

function simpleDebounce(fn, timeout = 0) {
  let timer;
  return () => {
    clearTimeout(timer);
    timer = setTimeout(fn, timeout);
  };
}

// debounced update runner function
const runAsyncForceUpdateQueueDebounced = simpleDebounce(
  runAsyncForceUpdateQueue
);

const inc = i => i + 1;

// Queus the async update and calles the debounced runner function
export default function useAsyncForceUpdate() {
  const [, setTick] = React.useState(0);
  const forceUpdate = React.useCallback(() => setTick(inc), []);
  const asyncForceUpdate = React.useCallback(() => {
    asyncForceUpdateQueue.push(forceUpdate);
    runAsyncForceUpdateQueueDebounced();
  }, [forceUpdate]);
  return asyncForceUpdate;
}
```

Near the entry point (root index.js or whatever) do something like this (or include patch file in webpack entry):
```js
import useAsyncForceUpdate from './useAsyncForceUpdate';
import { setDefaultForceUpdateHook } from 'mobx-react-lite';

setDefaultForceUpdateHook(useAsyncForceUpdate);

// Import rest of the application
import './app';
```

**Comment by FredyC** [comment]
Sorry, but I don't consider your "hotfix" something worth doing. It's not an immediate issue - nothing is broken, only warning is triggering, I would rather go for a proper solution instead. At least from your CodeSandbox I don't see that something would be broken.

> Yeah, the issue for me right now is that if I would like to apply this fix for the whole project, I would have to replace all imports of `observer` with a custom implementation.

You can just ignore the warning for now :)

> Instead what I would rather do is being able to set the default hook which should be used for `options.useForceUpdate` across all observer components.

Feel free to use [Yalc tool](https://github.com/whitecolor/yalc) or [patch-package](https://www.npmjs.com/package/patch-package) to locally fix the issue for you if warning is a big problem.

> I think about it, it makes sense that modifications to the observables should only cause rerenders **after** the current render finished.

That's arguable because it would mean double render. If we manage to "squeeze" that update in current render, it would be more efficient.

> Other than that I think the only real solution is to never use useAsObservableSource in components but instead pass everything you need as observable sources instead of props.

Yeah, that hook was created way before Concurrent mode was something to worry about and it did not cause any warnings before.

**Comment by olee** [comment]
> You can just ignore the warning for now :)
The problem is that it is not a warning but an error :)

> That's arguable because it would mean double render. If we manage to "squeeze" that update in current render, it would be more efficient.
Actually not - as observer components are memo components, they would only render once anyway (except if they also receive props). The only difference is that the second render happens **later** than right now.

And as I said before - just allowing users to set their own default forceUpdate scheduler (hook) is something I think should be possible anyway, like mobx allows configuring a reactionScheduler.
Basically the useForceUpdate hook for mobx-react-lite is the same as the reactionScheduler for mobx and that is why I definitely think this should be configurable.
Also this is only a very small change which would not impact any users at all.

**Comment by FredyC** [comment]
The `useForceUpdate` wasn't supposed to be exposed in the first place, it was done to support one library. What is your use case of overriding it besides hot fixing the warning? It shouldn't be needed in most cases.

**Comment by olee** [comment]
That's why I tried to explain the issue with reactionScheduler of mobx.

Not only does asynchronously running forceUpdate on components fix the update issue.
It could also be a performance improvement and possibly fix other issues, because changing different observable values during the rendering of a function component would cause multiple react rerenders whereas batching those with eg. react-dom's batchedUpdates would correctly order those updates and fix potential issues.

I agree that providing this default-option here is kinda a "crutch".
However the alternative would be to change the current default implementation of the forceUpdate in useObserver to be async by default.
But before doing this I think there are a lot of investigations & tests necessary.

The only thing that we know for sure is that at some point in the future many scenarios of using observer might break completely, because the react team said that those incorrect updates which generate this **error** (not warning) might not work at all in the future.

Therefore it really confuses my why you only see this as a "warning" but not as a critical issue.

**Comment by FredyC** [comment]
> Not only does asynchronously running forceUpdate on components fix the update issue.

I can agree with that, but that still doesn't provide a use case for exposing it. People shouldn't be implementing that by themselves.

> However the alternative would be to change the current default implementation of the forceUpdate in useObserver to be async by default.

I am totally ok with that as long as we make sure there are no memory leaks.

> Therefore it really confuses my why you only see this as a "warning" but not as a critical issue.

It is a warning right now, it doesn't break any actual code. We surely need to tackle it, but exposing crutches is not the way to go.

**Comment by olee** [comment]
So if I want to fix this for my project quickly, I will have to depend a fork I guess - well that's ok for me but I guess not everyone will want to do that.

>I am totally ok with that as long as we make sure there are no memory leaks.
I don't think memory leaks are a problem because all it does is delaying the forceUpdate call.

I tried patching the useAsyncForceUpdate implementation into mobx-react-lite and as expected, quite some tests will fail then.
**However**, most of these tests will work again by just moving the assertions into a `setTimeout(() => { ... })` so that the assertions run **after** the async updates have been applied (which would be correct then).
But as this might break some projects that relied on this weird behavior (should be few though I think), this might be a breaking change.

Also this would still require `setDefaultForceUpdateHook`-like configuration function, because the default implementation cannot rely on `react-dom` `unstable_batchedUpdates` (like batchingForReactDom). 
Alternatively we could provide the same kind of option like mobx (reactionScheduler) which would be used inside the implementation.
(Which is the reason why I still think it should be ok to merge the PR I created for now 🤔)

If you want to, I can try making a PR for this.

**Comment by FredyC** [comment]
> **However**, most of these tests will work again by just moving the assertions into a `setTimeout(() => { ... })` so that the assertions run **after** the async updates have been applied (which would be correct then).

That sounds like a terrible idea. Imagine how people are going to be writing tests for the actual app if such untracked change is happening there. But the `act` should help here if I am not mistaken.

> But as this might break some projects that relied on this weird behavior (should be few though I think), this might be a breaking change.

Yes, most likely. We not really ready for Concurrent mode in general, there are some existing issues to track that. Ideally the next major should tackle all that including this warning. But honestly, it still feels very far and nobody is paying that much attention to it since it's still experimental. It's totally possible it turns out that whole `useLocalStore/useAsObservableSource` idea is broken and it won't be possible to use that in the future.

> Also this would still require `setDefaultForceUpdateHook`-like configuration function, because the default implementation cannot rely on `react-dom` `unstable_batchedUpdates` (like batchingForReactDom).

Right, but we can just use currently exposed [batching scripts](https://github.com/mobxjs/mobx-react-lite/blob/master/batchingForReactDom.js), no need to have another function for that.

**Comment by olee** [comment]
I mean the problem is a general react state issue that affects not only libraries like mobx-react-lite and redux, but also react itself. Whenever state changes are required to happen during rendering, this issue will pop up.

For example
```jsx
function FailRedux() {
  if (someCondition) dispatch(someAction());
  return <div />;
}
```

Or even react itself:
```jsx
function Parent() {
  const [value, setValue] = React.useState(0);
  const ctx = React.useMemo(() => ({ value, update: setValue }), [value]);
  return (
    <MyCtx.Provider value={ctx}>
      <Child />
    </MyCtx.Provider>
  );
}

function Child () {
  const ctx = React.useContext(MyCtx);
  if (someCondition) ctx.update(10); // Will cause the react update error/warning
  return <div />;
}
```

> It's totally possible it turns out that whole useLocalStore/useAsObservableSource idea is broken and it won't be possible to use that in the future.
I do not think so as it is not really different from the react-only example above. I think it is just a matter of making stuff happen asynchronously which should also work fine with concurrent mode I think.
Basically **every** update that a user wants to perform during rendering (the `if (someCondition) doUpdate()`) **must** be deferred until after the current rendering process finished - I just think this is kinda a design philosophy of react.

That is also why I think it is actually correct that the checks in the tests have to be moved into a timeout or something like that.
It might of course be possible to somehow make those updates happen synchronously after the current rendering process, but before stuff like act() return like some kind of task queue that react runs automatically after the current rendering process finished (... this sounds just like useEffect I guess....).

I'm not sure if this is a good idea, but I think we should get in contact with some other library authors who might be affected like this etc. 
(redux? or maybe @mweststrate (sorry for pinging you if that's not ok with you))

**Comment by mweststrate** [comment]
I glanced at the problem, but the problem is too complicated to solve in 5
minutes and I don't have the headspace atm to investigate this deeply.
Let's revisit once mobx 6 is out of the door. I think we don't even need
that update / setState to happen if an update is already scheduled any way
cause of prop changes, it is just that we can't detect that case atm. But I
think we potentially could.

**Comment by mweststrate** [comment]
Generally speaking I really doubt that doing things async is the solution though. It is only suppressing the symptom, but not changing the behavior. It's just piling more complexity on top of the problem by introducing possibilities for race conditions, making everything depending on it downstream async, etc etc. So I'd really not recommend going in that direction.

**Comment by olee** [comment]
After working with it in a project for a while I can confirm that while it solves some issues, there are also problems caused by asynchronous dispatched updates.
In this case it was an input field which was updated.

When the update was asynchronous, react noticed a value change on the input component while the controlled value field still had the old one. That caused it to reset the input element value to the old one and only after the async render it would update to the new value. While this was not visible to the user directly, it cause the cursor to jump to the end of the input again.

I guess it's really necessary to either do this case-by-case, or find a completely different solution.
Or maybe in the end the react team themselves have to make it that calling setState from a different component is not considered an error any more.

**Comment by olee** [comment]
After some more thought, I actually found a solution that might be viable: https://github.com/olee/mobx-react-lite/commit/37fc829e871050d388d72fefa7106447f49d526b

The code change is definitely not perfect yet, but it solves the issue without breaking any of the tests and the solution is actually exactly what the react team always advices - to do state changes in useEffect.

It works by [intercepting reactions](https://github.com/olee/mobx-react-lite/blob/37fc829e871050d388d72fefa7106447f49d526b/src/observerBatcher.ts#L26-L38) while inside the [reaction.track]() in useObserver and moving their execution into a useEffect inside useObserver ([here](https://github.com/olee/mobx-react-lite/blob/37fc829e871050d388d72fefa7106447f49d526b/src/useObserver.ts#L134-L140)).

During my tests this solution proved to solve all the issues mentioned above and I could observe no negative side-effects.

What do you think about this solution?

Right now I'm not sure if it should be opt-out ([as now](https://github.com/olee/mobx-react-lite/blob/37fc829e871050d388d72fefa7106447f49d526b/src/useObserver.ts#L123)),opt-in or always active.
My main concern for always-active would be the useEffect running on every render, but I guess it should not be too much of a performance impact as mobx already reduces component rendering a lot.

**Comment by olee** [comment]
@FredyC did you have time to look into this yet? 😄

Also @mweststrate after testing my previous solution I totally have to agree, that doing it fully asynchronously is a failure, but pushing it at the end of the react render loop with useEffect seems to work wonders.

**Comment by FredyC** [comment]
@olee If you want to try to implement it in the repo with tests, it would be probably best. I won't have spare time in the near future, sorry.

**Comment by olee** [comment]
Well the PR is basically done - which is the commit I linked above.
As for tests, all the existing tests should pretty much cover this whole thing already.

However before actually going forward with a PR, I think this solution should be discussed in general.
The major point I am questioning is the way that the [reactions are intercepted](https://github.com/olee/mobx-react-lite/blob/37fc829e871050d388d72fefa7106447f49d526b/src/observerBatcher.ts#L26-L38) by using `reactionScheduler` from mobx.

If this solution is used, it would break once someone else tries to overwrite the mobx `reactionScheduler` with another implementation which I don't really like.
This is where @mweststrate comes into play: What we actually need would be a way to incercept reactions inside some block of code (like with track) so that their execution can be postponed without relying on the global `reactionScheduler`.

However, right now I don't see any other solution to solve this issue.

If you don't have time, maybe you know some other people who might be able to help out here?

Though if you're ok with it, I could try to refine this solution as it is now with some more tests and maybe adjust the docs etc. to contain a warning about changing reactionScheduler in mobx.

**Comment by mweststrate** [comment]
@olee if you want us to take a look at it, please open PR based on your fork, otherwise we're missing all the review tooling.

**Comment by urugator** [comment]
> However, right now I don't see any other solution to solve this issue.

Instead of posponing the reaction's effect with scheduler, can't we postpone the `forceUpdate` inside reaction's effect?:
```javascript
// useObserver

// Setup reaction:
const invalidateComponent = () => {
  // globals.effects is a reference to a queue of currently rendered component (if any)
  if (globals.effects) {
    // rendering - enqueue update for later
    globals.effects.push(() => forceUpdate()) 
  } {
    // not rendering - update immediately
    forceUpdate()
  }  
}
const newReaction = new Reaction("xxx", invalidateComponent)

// Prepare queue
const effects = [];
// Run updates previously enqueued during render
useLayoutEffect(() => {  
  effects.forEach(effect => effect())
  effects = []
})
// Render start
globals.effects = effects;
// Render
reaction.track(/* compacted */)
// Render end
globals.effects = undefined;
```
It's quite late, sorry if it's an utter nonsense.

**Comment by olee** [comment]
I thought of this as well but I thought it would not work, because the component which has to receive the forceUpdate might be a different one than the one that was rendered.

However from what I can see you circumvented this by introducing that global forceUpdate queue which then gets executed inside the currently rendering component's useEffect.

This looks really promising and would not require hijacking the reactionScheduler.
I could try writing up a PR with this solution or you could try it as well if you want to.

**Comment by mweststrate** [comment]
Out of curiosity, what is the real scenario in which the warning occurs. I
always thought as the hook solving issues with locally derived values (like
having a @computed field on a class component), not so much with shared
once. Not saying the problem doesn't need fixing, but want to understand
the bigger pic better first.

Op do 6 aug. 2020 00:53 schreef Björn Zeutzheim <notifications@github.com>:

**Comment by olee** [comment]
@mweststrate as for a mobx-react-lite related example you can check [my comment at the top](https://github.com/mobxjs/mobx-react-lite/issues/274#issuecomment-652958242) and [this comment](https://github.com/mobxjs/mobx-react-lite/issues/274#issuecomment-653524231) for a general demonstration of where this problem originates from.

To sum it up, react now says that calling setState of a different component from within the rendering of one component is regarded as an error.
It still works right now, but I guess they did this because it will fail in the future with react concurrent mode.

Right now this error will always occurr when the observable from `useAsObservableSource` is used in any other component than the current.
This is because inside the hook it will update the observable object and if any other component depends on one of those values, it will be forceUpdated via setState while still rendering the current component.

**Comment by urugator** [comment]
@mweststrate

> like having a @computed field on a class component

The thing is, in the class component nobody passes `this` to other components - therefore other components aren't force updated when `this.props` (the observable source) is imperatively mutated during render.
I think you could potentially hit the same problem if you would pass eg boxed computed that depends on `this.props`?

So yes, either the observable shouldn't be modified during render or it must not be passed to other components.
Forbidding to pass local store elsewhere is hard to force and probably reduces local store usefulness quite significantly.
Forbidding to modify observables during render means you cannot have this observable source (immediate conversion from non-observables to observables)

So the proposed workaround is - if forceUpdate is scheduled during render (due to an observable modifed during render) defer forceUpdate to `useLayoutEffect`, rather than calling it immediately during render.

**Comment by mweststrate** [comment]
Yeah, it is clear to me how the warning is caused :) It just makes me
wonder, is asAsObservableSource the right solution, or do we need a
different abstraction (e.g. a hoc, wrapper, dunno). That is why I'm
interested in more real life use cases like the router one, rather than
minimal repo's (which in turn are better to verify the solution). The risk
with any useEffect based solution is that it might easily cause double
renders. That in turn is probably fixable, just want to zoom out a bit
first and make sure, before we start piling fixes / workarounds, that we're
working in the right direction.

**Comment by urugator** [comment]
@mweststrate There is a double render already, currently a `setState` is scheduled during render, causing another one ... however in case of `setState` the react can probably prevent double rendering of other affected components, while `useEffect` will wait for their render too (not sure).

**Comment by mweststrate** [comment]
Yeah, the biggest problem is here indeed that unlike with classes, we don't have a `this` that can own an observable props object that is / can be updated outside render. Random thought: maybe it might be interesting to see if we can make an api where this `asObservabelSource` is basically pulled outside the render. E.g. 

```
export const Comp = observer(props => /* constract an observable object here */, observableProps => {
  // normal render
})
```

or make the props directly observable directly anyway in the HoC. There was a problem with that, but I don't remember anymore what exactly :) I'm going on holiday tomorrow, but I see if I can recall it after that.

**Comment by urugator** [comment]
Not sure if it helps aynthing, because it's the local store that is passable around and contains computeds that depends on this source (`observableProps`).
I think we shouldn't reduce the problem to observable props, which was mainly to allow the use of props in local computeds/reactions without causing staleness.
The goal here is to transform something non-observable to observable so it can be later used as such, even in different components. 
It's more akin to `getDerivedStateFromProps`, but you want to get state that is observable.
The analogy of `getDerivedStateFromProps` in functional components is calling `setState` during render (at a cost of double render and temporal inconsistentency of values), which is basically what we do (our state is synchronous, so no inconsistency).
The ideal situation would be: 
When invalidating currently rendered component call setState. 
When invalidating other component during render, defer the update to `useLayoutEffect`, but in case the component starts rendering before this layout effect is invoked, dequeue the update.
If an implementation is possible, it will be probably very fragile.

To be completely clear I don't like `useAsObservableSource/useLocalStore` I would get rid of them immediately, I am just trying to be constructive.

Enjoy your holiday :)

**Comment by urugator** [comment]
Another idea, hopefully completely avoiding double renders
```javascript
let isRendering = useRef(true);
let updateScheduled = useRef(false);

const reactionEffect = () => {
  if (globals.isRendering) {
    // something is rendering
    if (isRendering.current) {
      // this component is rendering
      // no need to do anything
    } else {
      // something else is rendering, schedule update
      updateScheduled.current = true;
    }
  } else {
    // not rendering anything, update immediately
    forceUpdate()    
  }
}
const reaction = new Reaction("xxx", reactionEffect)

useLayoutEffect(() => {
  if (updateScheduled.current) {
    forceUpdate()
  }
})

globals.isRendering = true;
isRendering.current = true;
updateScheduled.current = false;
reaction.track(/* render */)
isRendering.current = false;
globals.isRendering = false;
```

**Comment by olee** [comment]
> The risk
with any useEffect based solution is that it might easily cause double
renders

I think deferring the update into a useEffect will not have any impact, because afaik right now react internally does the same thing this patch does manually.
If you call setState on another component B from the currently rendering one A, the error is logged but component B is still queued to render after component A.
This fix would then result in the same amount of rendering and just follows the react semantics to properly call setState not during the rendering of another component.

> It just makes me
wonder, is asAsObservableSource the right solution, or do we need a
different abstraction

I think this could also be possible. Something like this maybe?

```tsx
import React from 'react';
import { observable } from 'mobx';

export interface WithObservablePropsProps<T extends Record<string, any>> {
    observableProps: T;
}

export default function withObservableProps<T extends Record<string, any>>(Component: React.ComponentType<WithObservablePropsProps<T>>) {
    const result: React.ComponentType<T> = function WithObservableProps(props: T) {
        const observableProps = React.useState(() => observable(props))[0];
        React.useEffect(() => {
            Object.assign(observableProps, props);
        });
        return <Component observableProps={observableProps} />;
    };
    return result;
}
```

And then using it like this:
```tsx
const MyComponent = observer(({ observableProps }: { observableProps: { num: number; str: string; }; }) {
    return (
        <div>{observableProps.num} / {observableProps.str}</div>
    );
});

const ComponentWithObservableProps = withObservableProps(MyComponent);

/*
type of ComponentWithObservableProps is
const ComponentWithObservableProps: React.FunctionComponent<{
    num: number;
    str: string;
}>
*/
```

However there are a few cons with this solution:
- You have to wrap your component (who still want's to use HOCs since hooks are here anyway??)
- Increases component nesting
- **It only works as a replacement for observable props.**
  The other solution would also solve the problem of modifying other stores passed from context etc. in the rendering function.
  eg stuff like running  `if (someCondition) myGlobalStore.prop = newValue;` inside a currently rendering component

**Comment by urugator** [comment]
> I think deferring the update into a useEffect will not have any impact, because afaik right now react internally does the same thing this patch does manually.

I would assume that react schedules setState immediately (at least for current component), while effects wait for the DOM, but really dunno.
But I agree the extra update in effect is fine, because there is simply no other place where we could schedule this update (and we can prevent updating components that were already rendered before effect)
Not sure if it can cause some ugly waterfall: effect -> render -> effect -> render -> effect.
However I am not convinced that the extra complexity is worth it - deriving observables from props should be an edge case (even more generally deriving **state** from props is an edge case). People should be discouraged from doing so and there are perhaps less optimal, but sufficient ways to handle this without `useAsObservableSource`.

> Something like this maybe?

We would have to move `memo` from `observer` to `withObservableProps` - or rather `withObservableProps` would use `useObserver`, therefore no extra nesting.
However I am more concerned about whether it's actually useful ... you can use these `observableProps` in `useLocalStore`'s `computed`, but that's pretty much it ... so you itroduce extra render with effect so you can safe another with computed.
I am actually worried people would start misusing it in all sorts of ways.

Another thing I am a bit worried: the `props` and `observableProps` are temporarily out-of-sync.
It doesn't seem to be a problem, because HOC only passes the "delayed" `observableProps`, but I think it can still cause some issues:
Eg the prop value could come from some observable store, then you obtain the store in Child component and the prop value doesn't match with the store value.
But to be fair I think something like this can already happen even without defering the update to effect.


Deleted the comment above, realized it can't work :D

**Comment by olee** [comment]
> I would assume that react schedules setState immediately (at least for current component), while effects wait for the DOM, but really dunno.

For the current one yes, but we are talking about other components - in this case from what I can infer, react currently schedules them to render after the current one.

> Not sure if it can cause some ugly waterfall: effect -> render -> effect -> render -> effect.

I think this might be possible, but I think it would only happen with code which would cause the same problem even with the current implementation (ie. modifying an observable value in a component during rendering which at the same time relies on this value):

```tsx
const store = observable({ value: 0 });
const FailComponent = observer(() => {
  store.value += 1;
  return <div>Value = {store.value}</div>;
});
```

> I am actually worried people would start misusing it in all sorts of ways.

As I said in my comment above - that this was just an example and this solution would have quite a few cons - so I also definitely would never use this solution.
I only posted it as an inspiration to think about so maybe someone else has a better idea based on this (though I don't really think there is any value in this).

> Another thing I am a bit worried: the props and observableProps are temporarily out-of-sync.

Would not be a problem, because the new values would be applied in a useEffect which would keep the changes still in sync in regards to the whole rendering process.

---

So to get back on topic: Are there so far any arguments against going forward with solution proposed in the PR #299 ?

**Comment by urugator** [comment]
> Would not be a problem, because the new values would be applied in a useEffect

It may break some assumptions in the render function ... what usually happens (to me) is `cannot read property of undefined`, because on the first render something isn't initialized and therefore must be accessed conditionally. But I can also imagine it breaking conditionally invoked effect, memo, etc.
Also you need `useLayoutEffect`, otherwise you could actually see the incorrect output flushed to real DOM.

> so far any arguments

My main argument is againts `useAsObservableSource`, which I have expressed above.
Then there is the mentioned problem with the use of `useAsObservableSource` outside of `observer`. Should we forbid it/warn (it would warn even in cases where react wouldn't...)?
And then I have a question about the concurrency (I really don't know how these fibers work):
Could it be that one fiber starts rendering, but the mutation notifies a reaction that belongs to a component of a different fiber, so the reaction would see there is a rendering in progress even though it's own fiber is not rendering at that moment? Does it makes sense, is it something to worry about?

**Comment by olee** [comment]
> It may break some assumptions in the render function

As I said before, that should not be possible.
Right now react itself schedules the forceUpdated component to render sometime after the current one while printing the error (ie. throws an error but still works).
The only difference is that we schedule the forceUpdate into the phase of react where it itself would rerender the component when forceUpdated incorrectly as in the current code.
That's why the rendering process itself should be pretty much exactly the same as before.

> Also you need `useLayoutEffect`

[Already using it](https://github.com/mobxjs/mobx-react-lite/pull/299/files?diff=split&w=1#diff-5c813a858d1bf62c0f14bf297cb96223R31)

> Then there is the mentioned problem with the use of useAsObservableSource outside of observer

I improved the PR to separate the delaying logic of force-update calls into a separate hook which can then be used in those places.
Also I think it would be useful in general to be able to access this utility hook externally so I added it to the exports.

**Comment by nulladdict** [comment]
Just a random thought I had, that might be relevant to the overall discussion.
So, initial problem is mixing observable and non-observable state, which results in tricky stuff we're dealing with right now.
Maybe, we should switch the approach from using non-observable state in computed variables and instead use observable state right in the render.

To illustrate, instead of this:

```jsx
const Component = observer(props => {
  const state = useLocalStore(source => ({ foo: 1 + source.bar }), props)
  /* ... */
})
```

We can encourage something like this:

```jsx
const Component = observer(props => {
  const state = useLocalStore(() => ({ foo: 1 }))
  const foo = state.foo + props.bar
  /* ... */
})
```

This way whatever uses non-observable state will be forced to compute synchronously right inside render, and we also might keep observable updates synchronous, unless there's outright mutation happening in render.
It certainly *feels* less cool and more verbose, but it's easier to reason about. If the computation is expensive, we can manually wrap it in `useMemo`

**Comment by urugator** [comment]
> As I said before, that should not be possible.

I was refering to that HOC solution, where the `Object.assign` is inside the effect, so the `observableProps` are not immediately synced.

> I think react already does this so this might not be necessary anymore.

Yea but it's not really obvious whats wrong, I quess we should at least mention it in docs.

> can you try to explain it again in other words?

Gave it some more thoughts, it's probably fine.

**Comment by olee** [comment]
@urugator kk I think then there are no more open questions.

Do you think we can then merge the PR?

**Comment by mweststrate** [comment]
Fixed in 2.1.0

**Comment by rizkyramadhan** [comment]
I still experienced this bug using nextjs and mobx-react-lite 2.2.0, maybe I did something wrong ?

**Comment by FredyC** [comment]
@rizkyramadhan Please open a new issue with all details and ideally minimal reproduction.
